donkeycar/templates/donkey2.py

```diff
@@ -132,48 +136,102 @@ def drive_mode(mode,
 
 
 
+def load_image(path):
+    img = Image.open(path)
+    return np.array(img)
+
+def get_generator(input_keys, output_keys, record_paths, meta, tub_path):
+    while True:
+        for record_path in record_paths:
+            with open(record_path, 'r') as record_file:
+                record = json.load(record_file)
+                inputs = [record[key] for key in input_keys]
+                outputs = [record[key] for key in output_keys]
+                input_types = [meta[key] for key in input_keys]
+                # output_types = [meta[key] for key in output_keys]
+                for i in range(len(inputs)):
+                    type = input_types[i]
+                    if (type == 'image_array'):
+                        inputs[i] = load_image("%s/%s" % (tub_path, inputs[i]))
+                    elif (type == 'custom/prev_image'):
+                        # Currently previous images are in array, but there is only one
+                        imagePath = inputs[i][0]
+                        inputs[i] = load_image("%s/%s" % (tub_path, imagePath))
+                yield inputs, outputs
+
+def get_batch_generator(input_keys, output_keys, records, meta, tub_path):
+    # Yield here a tuple (inputs, outputs)
+    # both having arrays with batch_size length like:
+    # 0: [input_1[batch_size],input_2[batch_size]]
+    # 1: [output_1[batch_size],output_2[batch_size]]
+    record_gen = get_generator(input_keys, output_keys, records, meta, tub_path)
+    while True:
+        raw_batch = [next(record_gen) for _ in range(cfg.BATCH_SIZE)]
+        inputs = [[] for _ in range(len(input_keys))]
+        outputs = [[] for _ in range(len(output_keys))]
+        for rec in raw_batch:
+            for i in range(len(input_keys)):
+                inputs[i].append(rec[0][i])
+            for i in range(len(output_keys)):
+                outputs[i].append(rec[1][i])
+        numpyInputs = [np.asarray(ar) for ar in inputs]
```

Comment:
Decide should we use _ or camelcase separator

Reply:
😔